scripts/main.py

```diff
@@ -0,0 +1,232 @@
+import os
+import time 
+import pickle
+import random
+import numpy as np
+import matplotlib.pyplot as plt
+import torch
+from multiprocessing import Pool
+import vboc.model as models
+from vboc.parser import Parameters, parse_args
+from vboc.abstract import SimDynamics
+from vboc.controller import ViabilityController
+from vboc.learning import NeuralNetwork, RegressionNN, plot_viability_kernel
+
+
+def computeDataOnBorder(N_guess, test_flag=0):
+    valid_data = np.empty((0, model.nx))
+    controller.resetHorizon(N_guess)
+
+    # Randomize the initial state
+    d = np.array([random.uniform(-1, 1) for _ in range(model.nv)])
+    q = np.array([
+        random.uniform(params.q_min + model.eps, params.q_max - model.eps)
+        for _ in range(model.nq)]
+    )
+
+    # Set the initial guess
+    x_guess = np.zeros((N_guess, model.nx))
+    u_guess = np.zeros((N_guess, model.nu))
+    x_guess[:, :model.nq] = np.full((N_guess, model.nq), q)
+
+    if not test_flag:
+        # Dense sampling near the joint position bounds
+        vel_dir = random.choice([-1, 1])
+        q_init = params.q_min + model.eps if vel_dir == -1 else params.q_max - model.eps
+        q_fin = params.q_max - model.eps if vel_dir == -1 else params.q_min + model.eps
+        i = random.choice(range(model.nq))
+        d[i] = random.random() * vel_dir
+        q[i] = q_init
+        x_guess[:, i] = np.linspace(q_init, q_fin, N_guess)
+
+    d /= np.linalg.norm(d)
+    controller.setGuess(x_guess, u_guess)
+
+    # Solve the OCP
+    x_star, u_star, N = controller.solveVBOC(q, d, N_guess, n=N_increment, repeat=50)
+    if x_star is None:
+        return None, None
+    if test_flag:
+        return x_star[0], x_star
+    # Add a final node
+    x_star = np.vstack([x_star, x_star[-1]])
+    # Save the initial state as valid data
+    valid_data = np.vstack([valid_data, x_star[0]])
+
+    # Generate unviable sample in the cost direction
+    x_tilde = np.full((N + 1, model.nx), None)
+    x_tilde[0] = np.copy(x_star[0])
+    x_tilde[0, model.nq:] -= model.eps * d
+
+    x_limit = True if model.checkVelocityBounds(x_tilde[0, model.nq:]) else False
+
+    # Iterate along the trajectory to verify the viability of the solution
+    for j in range(1, N):
+        if x_limit:
+            if model.checkStateBounds(x_star[j]):
+                x_limit = True
+            else:
+
+                if model.checkPositionBounds(x_star[j - 1, :model.nq]):
```

Comment:
Checking the position bounds on x_star should be useless, as x_star is the solution of the OCP, therefore it should always satisfy the position bounds.